spec/Type-selector.md

```diff
@@ -0,0 +1,127 @@
+# Candid Type Selector specification
+
+Date: Jun 10, 2024
+
+## Motivation
+
+We need a way to customize how we interpret Candid data for various use cases, for example,
+
+* Candid binding generation. One Candid type can map to several types in a host language, such as Rust and TypeScript. We may also want to adjust the type name and field name in the generated code to fit in the host language’s naming convention.
+* Pretty printer and Candid UI. One Candid type can have different meanings in different canisters. For example, a `blob` type can represent an image, an account id, a Wasm binary, etc. We want to customize the format when showing the data in the terminal or in the browser.
+* Data validation in deserialization. For canisters taking variable-length arguments, e.g., `blob`, `vec` and recursive types, we may want to limit the size of the input data to prevent spending too many cycles in deserializing invalid data.
+* Random value generator. We want to customize how we generate random Candid values for fuzz testing. For example, we may want to restrict the range of a `nat` type, or limit the maximal depth of a recursive type.
+
+We can of course implement all these customizations in each application, but it is beneficial if we can identify a common syntax to express the type level customization and apply the same syntax to all applications to enforce a consistent user experience and reduce the learning curve.
+
+## Design
+
+Motivated by Cascading Style Sheets ([CSS](https://en.wikipedia.org/wiki/CSS)), we define a configuration language to specify the customization of a Candid interface. The configuration file enables the separation of the Candid interface and its configuration, including layout, binding, validation, etc.
+
+There are two parts in the config language: 1) A way to quickly select particular type nodes in Candid types; 2) Attach semantic meaning to the selected type nodes.
+
+### Type selector
+
+For the first part, we define a type selector `<selector>` similar to CSS selector.
+
+```
+<selector> := <scopes> (. <path>)? | <path>
+<path> := <name> (. <name>)*
+<scopes> := <scope> (. <scope>)*
+<scope> := func:<name> | func:init | arg:<num> | ret:<num>
```

Comment:
Isn't this missing the `variant` and `record` selectors used in <paths>? As in `tree.variant.branch.record.left` below?

Reply:
Is this resolved? Or did I misunderstand something?

Reply:
Scope only refers to top level functions and arguments. `record` and `variant` is only part of the path. It should become clear when we get the full semantics specified out...